### PR TITLE
feat(federation): Add TalkV1 federation proxy for mentions

### PR DESCRIPTION
**Pull request by nickvergessen**
Decision is to proxy the requests on the own server,
so we can proxy other services (matrix coming to mind)
in the future and the clients still have to only talk to
their own server, to help with CSP and to reduce the efforts

### 🚧 Tasks

**Legend:**
- 🟥 Must have
- 🟨 Should have
- 🟩 Good to have
- ⚫ Not for now

---

- [x] Avatar
    - [x] getAvatar 🟨
    - [x] getAvatarDark 🟨
    - [x] ~~uploadAvatar~~ ⚫ Moderation is not supported in TalkV1
    - [x] ~~emojiAvatar~~ ⚫ Moderation is not supported in TalkV1
    - [x] ~~deleteAvatar~~ ⚫ Moderation is not supported in TalkV1
- [ ] Chat
    - [x] receiveMessages 🟥
    - [x] sendMessage 🟥
    - [x] deleteMessage 🟥
    - [x] editMessage 🟥
    - [x] getMessageContext 🟥
    - [ ] setReadMarker 🟥
    - [ ] setReminder 🟨 
    - [ ] getReminder 🟨 
    - [ ] deleteReminder 🟨 
    - [ ] markUnread 🟨 
    - [x] mentions 🟨 
    - [ ] shareObjectToChat 🟩
    - [ ] getObjectsSharedInRoomOverview 🟩
    - [ ] getObjectsSharedInRoom 🟩
    - [x] ~~clearHistory~~ ⚫ Moderation is not supported in TalkV1
- [ ] Room
    - [x] getParticipants 🟥
    - [x] 🆕 capabilities 🟥
    - [ ] setSessionState 🟨
    - [ ] …

### 🏁 Checklist

- [x] ⛑️ Tests (unit and/or integration) are included or not possible
- [ ] 📘 API documentation in `docs/` has been updated or is not required
- [ ] 🔖 Capability is added or not needed

### CODE REVIEW

File: lib/Controller/ChatController.php [hunk 1]
```diff
@@ -1154,6 +1157,12 @@ protected function getMessagesForRoom(array $messageIds): array {
 	#[RequirePermission(permission: RequirePermission::CHAT)]
 	#[RequireReadWriteConversation]
 	public function mentions(string $search, int $limit = 20, bool $includeStatus = false): DataResponse {
+		if ($this->room->getRemoteServer()) {
+			$proxy = \OCP\Server::get(ChatService::class);
```
Comment (nickvergessen) [inline lib/Controller/ChatController.php:1161]: FIXME

File: lib/Federation/CloudFederationProviderTalk.php [hunk 2]
```diff
@@ -120,6 +122,7 @@ public function shareReceived(ICloudFederationShare $share): string {
 		$remoteId = $share->getProviderId();
 		$roomToken = $share->getResourceName();
 		$roomName = $share->getProtocol()['roomName'];
+		$invitedCloudId = $share->getProtocol()['invitedCloudId'] ?? $this->cloudIdManager->getCloudId($shareWith, null);
```
Comment (nickvergessen) [inline lib/Federation/CloudFederationProviderTalk.php:125]: Should log when it was missing to help with debugging in the future

File: lib/Federation/Proxy/TalkV1/ChatService.php [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	) {
+	}
+
+	/**
+	 *
+	 * @return TalkChatMentionSuggestion[]
+	 */
+	public function mentions(Room $room, Participant $participant, string $search, int $limit = 20, bool $includeStatus = false): array {
+		$this->room = $room;
+		$headers = [
```
Comment (nickvergessen) [inline lib/Federation/Proxy/TalkV1/ChatService.php:54]: Should be abstracted away into a general proxy caller

File: lib/Federation/Proxy/TalkV1/ChatService.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+				'verify' => !$this->config->getSystemValueBool('sharing.federation.allowSelfSignedCertificates', false),
+				'nextcloud' => [
+					'allow_local_address' => $this->config->getSystemValueBool('allow_local_remote_servers'),
+				],
+				'headers' => $headers,
+				'timeout' => 5,
+				'auth' => [urlencode($participant->getAttendee()->getInvitedCloudId()), $participant->getAttendee()->getAccessToken()]
+			]);
+		} catch (\Exception $e) {
+			\OC::$server->getLogger()->error($e->getMessage(), ['exception' => $e]);
```
Comment (nickvergessen) [inline lib/Federation/Proxy/TalkV1/ChatService.php:77]: Inject logger and handle errors in a way the clients can be notified about it

File: lib/Federation/Proxy/TalkV1/ChatService.php [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+			// FIXME Error handling
+			return [];
+		}
+
+		$content = $response->getBody();
+		$data = json_decode($content, true);
+		return array_map([$this, 'flipLocalAndRemoteSuggestions'], $data['ocs']['data'] ?? []);
+	}
+
+	protected function flipLocalAndRemoteSuggestions(array $suggestion): array {
```
Comment (nickvergessen) [inline lib/Federation/Proxy/TalkV1/ChatService.php:87]: Missing reverse action (making a federated user of the own instance a local user)

File: tests/integration/features/federation/chat.feature [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    Given user "participant3" exists
+
+  Scenario: Get mention suggestions
+    Given the following "spreed" app config is set
+      | federation_enabled | yes |
+    Given user "participant1" creates room "room" (v4)
+      | roomType | 2 |
+      | roomName | room |
+    And user "participant1" adds federated_user "participant2" to room "room" with 200 (v4)
+    And user "participant1" adds user "participant3" to room "room" with 200 (v4)
```
Comment (nickvergessen) [inline tests/integration/features/federation/chat.feature:14]: add a test where p3 is also federated and then converted to a local one for p2

**Review by SystemKeeper — APPROVED** [review]

**Pull request merged**